src/openms/source/ANALYSIS/ID/IDMapper.cpp

```diff
@@ -299,107 +284,185 @@ namespace OpenMS
     // for statistics
     Size id_matches_none(0), id_matches_single(0), id_matches_multiple(0);
 
-    // iterate over the peptide IDs
+    // build map from file to peptide id
+//    std::map<String, std::unordered_map<String, const PeptideIdentification*>> file2nativeid2pepid;
+    bool has_spectrum_references{false};
     for (Size i = 0; i < ids.size(); ++i)
     {
-      if (ids[i].getHits().empty()) continue;
+      const PeptideIdentification* pid = &ids[i];
+
+      // skip IDs without peptide annotations
+      if (pid->getHits().empty()) continue;
+      String file_origin = pid->getMetaValue("file_origin", "");
+      String spectrum_reference = pid->getMetaValue(Constants::UserParam::SPECTRUM_REFERENCE);
+      // missing file origin is fine but we need a spectrum_reference if we want to build the map
+      if (spectrum_reference.empty()) continue;
+  
+      // TODO: check if there is already an entry
+//      file2nativeid2pepid[file_origin][spectrum_reference] = pid; // TODO: also use file information
+      has_spectrum_references = true;
+    }
 
-      getIDDetails_(ids[i], rt_pep, mz_values, charges);
+    std::unordered_map<String, ConsensusFeature*> nativeid2cf;
```

Comment:
Because logical separation in this code seems to be right now more important than 0-5% of speed gain for this linear task.

Reply:
Imho the logical separation is already given with the big if-case, no? All the mappings are even outside of the if-case so that they will be done needlessly in a non-TMT context.

Why not this:

```cpp
tmt = checkTMT(cmap); // additionally check the experiment_type in CMap
if (tmt == TMTMS2 || tmt == TMTMS3)
{
  warnForInvalidSettings();
  map = createFile2Scan2IDMap(ids);
  for (cf : cMap)
  {
    if (tmt == TMTMS2)
    {
        if (cf.hasMV(scan_id)) cf.attach(map.get(cf.getMV(scan_id)))
        else warnAndFail();
    } else if (tmt == TMTMS2) {
        if (cf.hasMV(id_scan_id)) cf.attach(map.get(cf.getMV(id_scan_id)))
        else warnAndFail();
    }
  }
} else {
  oldcode();
}
```

Reply:
I know what you mean... if wanted to do these warnAndFail checks already outside so I kept that code.
What I don't like that there is so much implicit stuff happening.
But I will take a look - the way you propose looks certainly easier to implement